pyfive/dataobjects.py

```diff
@@ -482,74 +487,131 @@ def find_msg_type(self, msg_type):
-            name = self.msg_data[offset:offset+name_size].decode(encoding)
-            offset += name_size
+        name_size_fmt = ["<B", "<H", "<I", "<Q"][flags & 3]
+        name_size = struct.unpack_from(name_size_fmt, data, offset)[0]
+        offset += size_of_length_of_link_name
 
+        name = data[offset:offset+name_size].decode(encoding)
+        offset += name_size
+
+        if dereference:
             if link_type == 0:
                 # hard link
-                address = struct.unpack_from('<Q', self.msg_data, offset)[0]
-                links[name] = address
+                address = struct.unpack_from('<Q', data, offset)[0]
             elif link_type == 1:
                 # soft link
-                length_of_soft_link_value = struct.unpack_from('<H', self.msg_data, offset)[0]
+                length_of_soft_link_value = struct.unpack_from('<H', data, offset)[0]
                 offset += 2
-                links[name] = self.msg_data[offset:offset+length_of_soft_link_value].decode(encoding)
-        return links
+                address = data[offset:offset+length_of_soft_link_value].decode(encoding)
+
+        return creationorder, (name, address)
+
+    def _iter_link_from_link_info_msg(self, info_msg):
+        """ Retrieve links from link info message. """
+        offset = info_msg['offset_to_message']
+        data = self._decode_link_info_msg(self.msg_data, offset)
+
+        heap_address = data["heap_address"]
+        name_btree_address = data["name_btree_address"]
+        order_btree_address = data.get("order_btree_address", None)
+        if name_btree_address:
+            yield from self._iter_links_btree_v2(name_btree_address, order_btree_address, heap_address)
+
+    def _iter_links_btree_v2(self, name_btree_address, order_btree_address, heap_address):
+        """ Retrieve links from symbol table message. """
+        heap = FractalHeap(self.fh, heap_address)
+        if order_btree_address:
+            btree = BTreeV2GroupOrders(self.fh, order_btree_address)
+        else:
+            btree = BTreeV2GroupNames(self.fh, name_btree_address)
+        adict = dict()
+        for record in btree.iter_records():
+            data = heap.get_data(record["heapid"])
+            creationorder, item = self._decode_link_msg(data, 0)
+            adict[creationorder] = item
+        for creationorder, value in sorted(adict.items()):
+            yield value
+
+    @staticmethod
+    def _decode_link_info_msg(data, offset):
+        version, flags = struct.unpack_from('<BB', data, offset)
+        assert version == 0
+        offset += 2
+        if flags & 1:
+            # creation order present
+            offset += 8
+
+        if flags & 2:
+            fmt = LINK_INFO_MSG2
+        else:
+            fmt = LINK_INFO_MSG1
+        data = _unpack_struct_from(fmt, data, offset)
+        return {k: None if v == 0xffffffffffffffff else v for k, v in data.items()}
```

Comment:
It's probably worthwhile to use UNDEFINED_ADDRESS where you're using 0xffffffffffffffff (it shows up in 3-4 places)

Reply:
Done